lahja/tools/benchmark/process.py

```diff
@@ -112,39 +122,41 @@ def launch(name: str, num_events: int) -> None:
             pass
 
     @staticmethod
-    async def worker(name: str, num_events: int) -> None:
-        event_bus = Endpoint()
-        await event_bus.start_serving(ConnectionConfig.from_name(name))
-        await event_bus.connect_to_endpoints(
-            ConnectionConfig.from_name(REPORTER_ENDPOINT)
-        )
+    async def worker(name: str, num_events: int, ready_event: Event) -> None:
+        with Endpoint() as event_bus:
+            await event_bus.start_serving(ConnectionConfig.from_name(name))
+            await event_bus.connect_to_endpoints(
+                ConnectionConfig.from_name(REPORTER_ENDPOINT)
+            )
 
-        events = ConsumerProcess.gen_with_timeout(
-            event_bus.stream(PerfMeasureEvent, num_events=num_events),
-            timeout=5
-        )
-        stats = LocalStatistic()
-        try:
-            async for event in events:
-                stats.add(RawMeasureEntry(
-                    sent_at=event.sent_at,
-                    received_at=time.time()
-                ))
-        except asyncio.TimeoutError:
-            print('Some messages from the consumer were dropped')
-
-        await event_bus.broadcast(
-            TotalRecordedEvent(stats.crunch(event_bus.name)),
-            BroadcastConfig(filter_endpoint=REPORTER_ENDPOINT)
-        )
-        event_bus.stop()
+            ready_event.set()
+
+            stats = LocalStatistic()
+            events = ConsumerProcess.gen_with_timeout(
+                event_bus.stream(PerfMeasureEvent, num_events=num_events),
+                timeout=5
+            )
+            try:
+                async for event in events:
+                    stats.add(RawMeasureEntry(
+                        sent_at=event.sent_at,
+                        received_at=time.time()
+                    ))
+            except asyncio.TimeoutError:
+                print('Some messages from the consumer were dropped')
```

Comment:
Answering to your comment https://github.com/ethereum/lahja/pull/43#discussion_r280908368 here.

>I'm starting to think that it's a bad idea to assume every message will be received, the sending process might start sending before the receiving process starts listening.

I think that our benchmark should ensure that we start sending messages only after the other process listens (which was the case so far) and that our benchmark should fail if messages get dropped. I believe accepting dropped messages in the benchmark could be dangerous because we may not notice a bug that would cause dropped messages for no plausible reason.

Reply:
Looks like, [this is the place](https://github.com/ethereum/lahja/pull/39/files#diff-d27046448bb81af24db4f2419116d9fcR69) that ensures the order in which the actions of the benchmark start is correct so save us from dropping events.

I'm much in favor of not sweeping dropped events under the rug then.

Reply:
Again, throwing solutions at problems I haven't really taken time to digest.  Using `multiprocessing.Event` in the processes which drive these seems like exactly the tool you're looking for...(don't broadcast until the receiving process has indicated it's ready by setting the event)

Reply:
> Using multiprocessing.Event in the processes which drive these seems like exactly the tool you're looking for...

Yeah, totally! That's actually what I was doing in this PR before the last change, which instead follows your suggestion of having the driver keep sending messages until the consumers have received enough.